### PR TITLE
Move cipher ctx 'original iv' parameter into the provider

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/evp/evp_lib.c [hunk 1]
```diff
@@ -140,11 +140,13 @@ int EVP_CIPHER_set_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
...
 
     if (type != NULL) {
+        oiv = (unsigned char *)EVP_CIPHER_CTX_original_iv(c);
         j = EVP_CIPHER_CTX_iv_length(c);
         OPENSSL_assert(j <= sizeof(c->iv));
-        i = ASN1_TYPE_set_octetstring(type, c->oiv, j);
+        i = ASN1_TYPE_set_octetstring(type, oiv, j);
     }
     return i;
 }
```
[the remark this answers is not in the document]
Comment (levitte) [reply crypto/evp/evp_lib.c]: Don't get me wrong, it's a neat trick...  but quite frankly, if *I* would code a provider, the thought of handing over a pointer to libcrypto where I know that I'm letting go of all control has me go kinda [eeeep]

[the remark this answers is not in the document]
Comment (slontis) [reply crypto/evp/evp_lib.c]: Err....
Why is this any different to what EVP_CIPHER_CTX_iv()  is doing?

[the remark this answers is not in the document]
Comment (slontis) [reply crypto/evp/evp_lib.c]: ASN1_TYPE_set_octetstring() makes a copy of the data..

[the remark this answers is not in the document]
Comment (slontis) [reply crypto/evp/evp_lib.c]: We could also use an octect_string instead of a pointer and copy into the evp's ctx->oiv if that makes you happier..

[the remark this answers is not in the document]
Comment (levitte) [reply crypto/evp/evp_lib.c]: Argh!  Ignore what I said...  I mixed this up with what `EVP_CIPHER_get_asn1_iv` does...

**Comment by levitte** [comment]
Naming wise, I wonder if the parameter isn't a bit misnamed.  What does "original" mean?  What origin does it refer to?  So I wonder if it wouldn't be better to call it "initial IV", to express that it's the IV that was specified when initializing the cipher, i.e. when calling `EVP_CipherInit` & friends.

(remember that `EVP_CipherInit` can be called more than once on the same context, so "original" would make me wonder if it's the IV from the first `EVP_CipherInit` on the same context, or the last, if I didn't know the code as well as I do)

File: doc/man7/provider-cipher.pod [hunk 2]
```diff
@@ -235,7 +235,11 @@ Gets the IV length for the associated cipher algorithm.
 
 =item B<OSSL_CIPHER_PARAM_IV> (octet_string OR octet_ptr)
 
-Gets the IV for the associated cipher ctx.
+Gets the current IV for the associated cipher ctx.
+
+=item B<OSSL_CIPHER_PARAM_ORIGINAL_IV> ("oiv") <octet_string OR octet_ptr>
```
Comment (levitte) [inline doc/man7/provider-cipher.pod]: "oiv" isn't the best of names.  "original-iv" or "initial-iv" would be better, which one is a matter of debate...

Comment (levitte) [reply doc/man7/provider-cipher.pod]: Also, we've changed the parameter naming convention to this:

```
=item "name" (B<OSSL_BLAH_PARAM_NAME>) <type spec>
```

**Comment by levitte** [comment]
Speaking of IV, I'm starting to wonder why we even share the *running* IV (which, strictly speaking, isn't an **I**V any more) from providers.  With *legacy* code, I fully understand, because all back-ends needed to access the IV that was centrally stored in the `EVP_CIPHER_CTX` instead of their own private context.  With provided implementations that hold the IV in the own private context, there should be about zero need to share that one, just the *actual* initialization vector.

In other words, there should really just be a `OSSL_CIPHER_PARAM_IV`, and it should only access what we currently call "the original IV".

Perhaps this is a matter for another PR.  I'll be happy to oblige.

**Comment by beldmit** [comment]
I need to look at it more thoroughly. GOST TLS requires an original IV as a source to derive the per-packet IV. See https://tools.ietf.org/html/draft-smyshlyaev-tls12-gost-suites-05#section-4.1.1 for details.

File: doc/man7/provider-cipher.pod (hunk 2 above)
Comment (slontis) [reply doc/man7/provider-cipher.pod]: Removed this from the pod file - it now just uses OSSL_CIPHER_PARAM_IV

**Comment by slontis** [comment]
> there should really just be a `OSSL_CIPHER_PARAM_IV`, and it should only access what we currently call "the original IV".

Good point - If there is a need for the running IV (which I dont think there is now) we would add it later.
I have removed the references to the original iv param - and it  seems to work fine.

**Comment by slontis** [comment]
>  need to look at it more thoroughly. GOST TLS requires an original IV as a source to derive the per-packet IV. See https://tools.ietf.org/html/draft-smyshlyaev-tls12-gost-suites-05#section-4.1.1 for details.

TLS related ciphers normally use a seperate mechanism via some ctrls to set the fixed_iv and aad so this should not be an issue.

File: providers/common/ciphers/cipher_common.c [hunk 3]
```diff
@@ -387,6 +383,21 @@ int cipher_generic_set_ctx_params(void *vctx, const OSSL_PARAM params[])
...
+int cipher_generic_initiv(PROV_CIPHER_CTX *ctx, const unsigned char *iv,
+                          size_t ivlen)
+{
+    if (ivlen != ctx->ivlen
+        || ivlen > sizeof(ctx->iv)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_IVLEN);
+        return 0;
+    }
+    ctx->iv_set = 1;
+    ctx->ivlen = ivlen;
```
[the remark this answers is not in the document]
Comment (levitte) [reply providers/common/ciphers/cipher_common.c]: (as far as I can read, you *shouldn't* set `ctx->ivlen` here either way, because that carries information of what the IV length *should* be, according to information you have from algorithm setup, right?)

File: providers/common/ciphers/cipher_gcm.c [hunk 4]
```diff
@@ -56,7 +51,7 @@ static int gcm_init(void *vctx, const unsigned char *key, size_t keylen,
             return 0;
         }
         ctx->ivlen = ivlen;
-        memcpy(ctx->iv, iv, ctx->ivlen);
+        memcpy(ctx->iv, iv, ivlen);
```
[the remark this answers is not in the document]
Comment (slontis) [reply providers/common/ciphers/cipher_gcm.c]: The AEAD ciphers use special case structs that do not use the base CTX that is used by the cipher_generic_iv.. I could change them to do that but then it gets a lot nastier when there are 3 layers of structs (2 is bad enough).

File: providers/common/ciphers/cipher_common.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (slontis) [reply providers/common/ciphers/cipher_common.c]: I will remove the line - but the question is not so straightforward.. it is really set up with a default iv length - if this could be changed then it would make sense (but that is not how we are caliing it).

**Comment by slontis** [comment]
Updated

File: providers/common/ciphers/cipher_gcm.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (levitte) [reply providers/common/ciphers/cipher_gcm.c]: Okie, no biggie, just curious

[the remark this answers is not in the document]
Comment (levitte) [reply providers/common/ciphers/cipher_gcm.c]: Oh, however, shouldn't you set `ctx->iv_set`?

[the remark this answers is not in the document]
Comment (levitte) [reply providers/common/ciphers/cipher_gcm.c]: (I've a hard time seeing if `ctx->iv_set` is used in the GCM case...)

[the remark this answers is not in the document]
Comment (slontis) [reply providers/common/ciphers/cipher_gcm.c]: it doesnt use iv_set as the iv setup is more complex for AEAD.. You can see however that it sets iv_state (which is the equivalent)

[the remark this answers is not in the document]
Comment (levitte) [reply providers/common/ciphers/cipher_gcm.c]: ... it's a bit confusing.  Almost cryptic, which is appropriate, considering... :wink:

[the remark this answers is not in the document]
Comment (slontis) [reply providers/common/ciphers/cipher_gcm.c]: not really - they are using completely different contexts that dont share the same base ctx. So the fields are different. 
gcm sets the iv using quite a few different paths as well as generating the iv internally, so it is more complex that just the one path that the normal ciphers use to just init the iv.

[the remark this answers is not in the document]
Comment (levitte) [reply providers/common/ciphers/cipher_gcm.c]: Ah, right

**Comment by slontis** [comment]
rebased due to simple merge conflict in header ciphercommon.h. Approval should still apply.

**Comment by levitte** [comment]
> Approval should still apply.

Not formally.  A re-review will confirm that you didn't goof up the merge fixups

**Comment by slontis** [comment]
Thanks.. Merged to master.
